protocols/modbus/src/main/resources/protocols/modbus/modbus.mspec

```diff
@@ -247,9 +247,9 @@
 ]
 
 [type 'ModbusPDUReadFileRecordResponseItem'
-    [implicit   uint 8     'dataLength'     '(COUNT(data) * 2) + 1']
+    [simple   uint 8     'dataLength']
```

Comment:
I think in this case the field could remain an implicit ... just change the syntax to "(COUNT(data) + 1)" cause this way you don't have to manually manage it.

Reply:
Hi Chris,

For this response you can have multiple sub responses/response items within the one response. In the overall PDU/response header there is only a total length and function code, there isn't any details about the length of each sub response or the number of sub responses. 
I don't see a way of determining the number of sub responses without using the length from each sub response.

Reply:
But if there is no 100% link between the dataLength and the length of the data ... we will be having trouble during the parsing of the messages.

Do I understand this correctly, that if in the "data" part there are 3 short items then you want to set this to 3 instead of 6? If the problem is that you need to know the request in order to decode the response, you can treat the response as a raw byte-array payload and split things up in the driver. For S7 I don't know how to decode the response unless I know the request ... same with the KNX protocol, where you even need to pull in some external data structures in order to know how to decode the payload ... It's not that uncommon.

Reply:
Hi Chris,

The response does include the length of the sub-response within each sub-response, but this length can't be implied without using the other information such as from the request. Changing it to simple uses the length returned within the response.

The length included in the sub-response is in bytes (not including the datalength byte), so if three short items are being returned the length would be 6 + 1.

Reply:
Woudln't this exactly match my proposed: "(COUNT(data) + 1)"?

Reply:
Yes the value works out the same, but if you used implicit then you assume (Without using the returned value) that you know the length of the data in the response. 

If we take an example with a response like this, we queried 2 memory areas. (in our case this would only happen when we request data that spans a file record e.g.609999[2]). The response would be:-

0x14 - Function Code
0x08 - Length of Overall Response
0x03 - Length of Sub Response 1
0x06 - Reference Type
0x00 - Data Byte 1
0x01 - Data Byte 2
0x03 - Length of Sub Response 2
0x06 - Reference Type
0x00 - Data Byte 1
0x00 - Data Byte 2

If we used implicit for the sub response length, when parsing sub response 1 we would assume the rest of the overall response is the data section in sub response 1 and the length would be 7 instead of 3. We would then not parse the second sub response.

Reply:
Hi,

sorry for keeping on asking ... but I still don't seem to grasp the problem ... so translating this to objects from the mspec. In your case you would have:

0x14 -> error = false & function = 0x14
- ModbusPDUReadFileRecordResponse
    0x08 -> byteCount (Used to know how many ModbusPDUReadFileRecordResponseItems have to be parsed)
    - ModbusPDUReadFileRecordResponseItem[0]
        0x03 -> dataLength (used to know the size of the byte array, which is 3 - 1 = 2)
        0x06 -> referenceTypte = 6
        0x00 -> recordData[0]
        0x01 -> recordData[1]
    - ModbusPDUReadFileRecordResponseItem[1]
        0x03 -> dataLength (used to know the size of the byte array, which is 3 - 1 = 2)
        0x06 -> referenceTypte = 6
        0x00 -> recordData[0]
        0x01 -> recordData[1]

I can't see the problem here ... the byteCount of ModbusPDUReadFileRecordResponse matches the number of bytes consumed by all ModbusPDUReadFileRecordResponseItems together and the dataLength in each ModbusPDUReadFileRecordResponseItem matches the length of the bytes in the data array + the one byte for the "referenceType".

Reply:
Hi Chris,

Maybe I'm confused by what implicit and simple mean. I take them to mean:-
Simple - Use the value that is included in the response.
Implicit - Attempt to calculated the value based on the other fields in the response.

As the length of the ModbusPDUReadFileRecordResponseItems is included in the response I changed it to simple instead of trying to calculate it from the data field.

Reply:
Also as the data field has a variable length if you try to infer the length of it, it wouldn't know the length of the data field in the first ModbusPDUReadFileRecordResponseItem.

So when it starts parsing the first ModbusPDUReadFileRecordResponseItem it would receive:-
0x03 - It would then try to infer this field by using the rest of the message. It would end up with 7 instead of 3.
0x06
0x00
0x01
0x03
0x06
0x00
0x01

Reply:
No ... it would read the 0x03 as length (When parsing implicits are parsed just like simple fields, however they are not stored in a property as this information can be calculated by the given expression when serializing) ... so when parsing the 0x03 would be parsed and would be available to other parsing expressions as dataLength (which would be used to know how many elements of "data" have to be read). When serializing instead of reading a property of the object, it would check how many elements the "data" array has and add 1.

Reply:
Oops sorry about that, I shouldn't assume things.

Your correction works.